### PR TITLE
Add support for dual ADC mode.

### LINKED ISSUE #54: Synchronized ADC Capture
`AdvancedADC` gives you two options for capturing two analog signals simultaneously.  The first is to bind both analog input pins to the same `AdvancedADC` instance.  This works great, but sadly, means your sample rate is cut in half. The `sample_rate` provided applies to the sampling of both inputs. So a 1M Samp./sec rate really means each is sampled at 500K Samp./sec. If you trying to push the limits of sample rate, this is not a great solution.

The second way is to bind each analog input to a different `AdvancedADC`. The library will automatically assign the two pins to two different ADC blocks. The sample rate of each will be maximum.  Very nice.  However, this creates another problem.  The `begin()` routine takes approximately 100 us on GIGA R1 to initialize and start the ADC capture.  So if I say something like:

               
```cpp
adc_input[0].load(res, sample_rate, num_samples, QUEUE_DEPTH,1,&(probes[0]));
adc_input[1].load(res, sample_rate, num_samples, QUEUE_DEPTH,1,&(probes[1])));
```

What happens is that the second ADC input capture is delayed by about 100us.  So the two signals are not synchronized.  If you're trying to capture two signals with reasonable synchronization, that is not going to work.  I analyzed the timing of each line of code in the `begin()` routine and found that two lines are causing almost all the delay:

https://github.com/arduino-libraries/Arduino_AdvancedAnalog/blob/f7d081c4a56f1a87f8b4a9f027f2efb2b4174b7c/src/AdvancedADC.cpp#L187

and

https://github.com/arduino-libraries/Arduino_AdvancedAnalog/blob/f7d081c4a56f1a87f8b4a9f027f2efb2b4174b7c/src/AdvancedADC.cpp#L200-L202

I have a solution.  I created two new routines in `AdvancedADC.cpp`  I call `load()` and `fire()`.  
`load()` is identical to the `begin()` routine up to a point, including the long-delay lines above.  It then returns.  Essentially `load()` allocates the DMA and ADC etc, but does not start the capture.  The `fire()` routine (which only takes about 4 us) does the rest of
what was in `begin()` and starts the capture.  This is working and providing a synchronization error of about 4-6 us.  Not bad. 

This is probably NOT the solution that someone else would prefer, but it works for me.  Now I have full sample-rate and okay synchronization between two or three analog inputs.

I guess I would say  the issue I have is that `begin()` takes too long (106 us!).  Either this needs to be shortened significantly or the idea of the `load()` and `fire()` would also work.

### LINKED ISSUE #59: Added dual-mode support for ADC.
After reading the 3500-page reference manual (nicely written but so much detail...so much...) for the STM32H7xx
I found how to enable the dual-mode.

I modified the begin() and start() routines and added a few new members to the AdvancedADC class to allow using dual-mode.
This gives the optimal synchronization between two sampled inputs.  I'm seeing between 0 and 1us skew, which is pretty good for a 1M samp/sec configuration.

It does have some restrictions.  You need to always use ADC1 and 2 (dual mode is only supported in HW for ADC 1 and 2).  Also, you need to configure the AdvancedADC instance to explicitly selected dual mode.  I tried to keep it as general purpose as possible.  

I would be happy to share my code if you're interested.    I can do a pull_request.

### PR DESCRIPTION
**Pull request by iabdalkader**
Fixes #54
Fixes #59

### CODE REVIEW

**Comment by iabdalkader** [comment]
I don't think we need to do that, see the update API.

**Comment by iabdalkader** [comment]
We can't assign pins dynamically with this, but I don't think we need to support this for dual ADCs.

**Comment by jmdodd95682** [comment]
Yes.  You could make it a requirement that pins are assigned upfront to each instance of AdvancedADC prior to calling the AdvancedADCDual.begin().  This solves the problem as well.  I'll need to make a few minor changes to my code because I am calling it incorrectly.  I'll make the changes to my app and see if everything works.

**Comment by iabdalkader** [comment]
No problem, take your time. Note there's an example here, but if you have a more useful one please send it.

**Comment by jmdodd95682** [comment]
I'm having a real problem with the refactored code.  I'm finding it difficult to declare an instance of AdvancedADCDual as a member of another class.  Since there is no blank constructor I have to declare it as a pointer of type AdvancedADCDual*.  A snippet of my code is below.

`class osc{
    private:

        uint16_t sample_period;
        uint8_t probes[MAX_NUM_PROBES];
        uint16_t num_samples;
        uint32_t sample_rate;
        bool    dual_mode;
        uint32_t timestamp;
        uint32_t sampleDelay;
        
    public:
        uint32_t max_num_pins;
        AdvancedADC adc_input[2];
        AdvancedADCDual *adc_dual;

        osc(uint8_t *p,uint8_t ena, uint16_t ns,uint32_t sr) 
        {
            max_num_pins=MAX_NUM_PROBES;
            for(uint8_t i=0;i<MAX_NUM_PROBES;i++)
            {
                probes[i]=p[i];
                adc_input[i]=AdvancedADC(p[i]);
            }
            adc_dual=new AdvancedADCDual(adc_input[0],adc_input[1]);`

It compiles with no errors.  Calling the begin() also seems to work correctly as far as I can tell, but when I go to read() the captured data its like ADC is not actually running.  I never receive any data back from the dma buffers.

I'm going to try running your example and see if that works for me.

**Comment by jmdodd95682** [comment]
Your example works fine.  When I globally declare the AdvancedADC instances and the AdvancedADCDual instance in a similar way my app also works fine as well.  So, I must have some problem with how I'm declaring or using the AdvancedADCDual.  

So, I would say the refactored code is fine.  I've just got to get a little more clever about how I use the API.

File: src/AdvancedADC.cpp [hunk 1]
```diff
@@ -212,27 +229,82 @@ int AdvancedADC::begin(uint32_t resolution, uint32_t sample_rate, size_t n_sampl
     hal_dma_enable_dbm(&descr->dma, descr->dmabuf[0]->data(), descr->dmabuf[1]->data());
     HAL_NVIC_EnableIRQ(descr->dma_irqn);
 
-    // Init, config and start the ADC timer.
+    if (start) {
+        return this->start(sample_rate);
+    }
+
+    return id();
```
Comment (leonardocavagnis) [inline src/AdvancedADC.cpp:236]:
```suggestion
    return 1;
```

File: src/AdvancedADC.cpp [hunk 2]
```diff
@@ -212,27 +229,82 @@ int AdvancedADC::begin(uint32_t resolution, uint32_t sample_rate, size_t n_sampl
...
 
+int AdvancedADCDual::begin(uint32_t resolution, uint32_t sample_rate, size_t n_samples, size_t n_buffers) {
+    // The two ADCs must have the same number of channels.
+    if (adc1.channels() != adc2.channels()) {
+        return 0;
+    }
+
+    // Configure the ADCs.
+    // Note only ADC1 (master) and ADC2 can be used in dual mode.
+    if (adc1.begin(resolution, sample_rate, n_samples, n_buffers, false) != 1) {
```
Comment (leonardocavagnis) [inline src/AdvancedADC.cpp:279]:
```suggestion
    if (!adc1.begin(resolution, sample_rate, n_samples, n_buffers, false)) {
```

File: src/AdvancedADC.cpp [hunk 3]
```diff
@@ -212,27 +229,82 @@ int AdvancedADC::begin(uint32_t resolution, uint32_t sample_rate, size_t n_sampl
...
+        return 0;
+    }
+
+    if (adc2.begin(resolution, sample_rate, n_samples, n_buffers, false) != 2) {
```
Comment (leonardocavagnis) [inline src/AdvancedADC.cpp:283]:
```suggestion
    if (!adc2.begin(resolution, sample_rate, n_samples, n_buffers, false)) {
```

File: src/AdvancedADC.cpp [hunk 4]
```diff
@@ -212,27 +229,82 @@ int AdvancedADC::begin(uint32_t resolution, uint32_t sample_rate, size_t n_sampl
...
+        return 0;
+    }
+
```
Comment (leonardocavagnis) [inline src/AdvancedADC.cpp:286]:
```suggestion

// Note only ADC1 (master) and ADC2 can be used in dual mode.
if (adc1.id() != 1 || adc2.id() != 2) {
    return 0;
}

```

File: src/AdvancedADC.cpp [hunk 5]
```diff
@@ -212,27 +229,82 @@ int AdvancedADC::begin(uint32_t resolution, uint32_t sample_rate, size_t n_sampl
...
 }
 
+int AdvancedADCDual::begin(uint32_t resolution, uint32_t sample_rate, size_t n_samples, size_t n_buffers) {
+    // The two ADCs must have the same number of channels.
+    if (adc1.channels() != adc2.channels()) {
+        return 0;
+    }
+
+    // Configure the ADCs.
+    // Note only ADC1 (master) and ADC2 can be used in dual mode.
```
Comment (leonardocavagnis) [inline src/AdvancedADC.cpp:278]:
```suggestion
```

**Comment by iabdalkader** [comment]
@leonardocavagnis updated.

**Review by leonardocavagnis — APPROVED** [review]

**Pull request merged**